### DISCUSSION TITLE
npm install should use publishConfig.registry when used with a package.json

### DISCUSSION THREAD

**Discussion by ARitz-Cracker** [npm/feedback#131] [CLI] [answered]
I decided to start this here instead of submitting an issue with npm as I wouldn't necessarily call it a "bug".
I believe that when I use `npm install` inside a project with a `package.json`, it should respect `publishConfig.registry`, the same way `npm publish` does.

Of course I could change my default registry, this is also true for publishing. However, since I have the option to change my publish registry on a package-by-package basis in order to ensure everything is sent where it needs to go, it makes sense that the packages themselves would expect dependencies from the registry they are published to, which makes explicitly stating the registry on install a bit redundant.

I assume the creation of the `publishConfig.registry` option was created to avoid accidentally publishing with the default registry, there _can_ be consequences which come from accidentally installing using the default registry as well. A completely different package with the same name as the dependency which the app wasn't expecting could be installed, which would result in application failure at best and malicious targeted attacks at worst. This could be mitigated by creating an organization with 0 packages and publishing anything in the private registry to that scope, reducing the potential severity to mild annoyance and the leak of a private packages name, but, again, the existence of a package property which allows you to set the registry sets the expectation that this isn't the kind of issue you'd have to think about.

Let me know what you think!

**Comment by MylesBorins** [comment, marked as answer]
hey @ARitz-Cracker ... I'd personally be worried about some of the ways a feature like this could fragment / fracture the ecosystem. For example... if I work for a company and I'm using a private registry behind a firewall, I want to make sure all package installs are going to that registry. I would not want individual packages to be able to specify that they call out somewhere else.

I could imagine that a feature like this might be useful at a project level, not package level, but I feel like creating yet another place to configure this could also be problematic.

pinging @npm/cli to verify my gut feel here.

**Reply by ARitz-Cracker** [reply]
I understand where you were coming from, which is why I chose not to respond further until now.

However, it would appear my suggestion could have prevented [remote code execution exploits in development pipelines](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610). I agree that having these organizations configure everything properly from the get-go would also prevented this, (for example, I always used scoped package names on all my self-hosted packages) but, for better or for worse, we're relying more and more on our tools catching these kinds of mistakes for us, especially when, again, the existence of a property called "registry", which _already_ changes NPMs behaviour, creates the reasonable assumption that what's specified would apply to installation as well.

**Comment by thw0rted** [comment]
This is what a `.npmrc` file is for: NPM configuration settings that should be applied to all commands executed in the given directory.  `publishConfig` is an additional set of options that should be applied only during `npm publish`, hence the name.  Applying `publishConfig` when you're not, y'know, *publishing*, would be a Principle Of Least Surprise violation.

Your linked article is about a failure due to a number of compounded errors.  *Obviously* companies should only publish proprietary libraries under a scope, that's why scopes exist.  Even if the company had an internal package mirror set up, using that registry exclusively would not have fixed the failure to scope correctly.  Paypal could publish the package `pp-internal-thing@1.0.0` to their internal registry, but as long as it's configured to pull from public `npmjs.org` as an upstream, and the Paypal devs didn't think to reserve the package name on public NPM, hackers could squat on `pp-internal-thing` and publish version 1.999.0.  The internal mirror would likely fetch that package during a routine refresh and CI pipelines would happily install it.  The only fix for this is to actually use scopes correctly as they were designed, setting the internal registry in the project `.npmrc`.
